src/EditorFeatures/Core/Classification/Syntactic/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -313,29 +326,29 @@ static ITextSnapshot GetLatest(ImmutableSegmentedList<ITextSnapshot> snapshots)
                     return latest;
                 }
 
-                async ValueTask<SnapshotSpan> ComputeChangedSpanAsync()
+                static async ValueTask<SnapshotSpan> ComputeChangedSpanAsync(SolutionServices solutionServices, IClassificationService classificationService, SumType<SyntaxNode, Document>? previousDocumentOrRoot, ITextSnapshot currentSnapshot, Document currentDocument, SyntaxNode? currentRoot, TimeSpan diffTimeout, CancellationToken cancellationToken)
                 {
-                    var changeRange = await ComputeChangedRangeAsync().ConfigureAwait(false);
+                    var changeRange = await ComputeChangedRangeAsync(solutionServices, classificationService, previousDocumentOrRoot, currentDocument, currentRoot, diffTimeout, cancellationToken).ConfigureAwait(false);
                     return changeRange != null
                         ? currentSnapshot.GetSpan(changeRange.Value.Span.Start, changeRange.Value.NewLength)
                         : currentSnapshot.GetFullSpan();
                 }
 
-                ValueTask<TextChangeRange?> ComputeChangedRangeAsync()
+                static ValueTask<TextChangeRange?> ComputeChangedRangeAsync(SolutionServices solutionServices, IClassificationService classificationService, SumType<SyntaxNode, Document>? previousDocumentOrRoot, Document currentDocument, SyntaxNode? currentRoot, TimeSpan diffTimeout, CancellationToken cancellationToken)
                 {
                     // If we have syntax available fast path the change computation without async or blocking.
-                    if (previousRoot != null && currentRoot != null)
-                        return new(classificationService.ComputeSyntacticChangeRange(currentDocument.Project.Solution.Services, previousRoot, currentRoot, _diffTimeout, cancellationToken));
+                    if ((previousDocumentOrRoot?.TryGetFirst(out var previousRoot) ?? false) && currentRoot != null)
```

Comment:
honestly, given the need to both null check, and then try to get a value, and then ?? it, it feels like this would be *much* simpler with just `(Document? previousDocument, SyntaxNode? previousRoot)` (where the tuple is non-null itself).

Reply:
The previous type didn't represent the either/or state correctly. The current type accurately presents the following as the complete set of valid states:

* Last snapshot and syntax root are not null, but document is null
* Last snapshot and document are not null, but syntax root is null
* Last snapshot and document and syntax root are all null

I'm open to simplifying refactorings provided they retain this static representation of correctness.

Reply:
i'll look when non-static.  it may be fine at that point :)

Reply:
much nicer, though i still hate the `?. ... ?? false` pattern.  *consider* having a TryGetFirst/Second extension on `SumType<>?` to make this much simpler/clearer.

Reply:
Was able to rewrite this to avoid the `?.` and `??`.